lib/blocks.php

```diff
@@ -148,119 +148,61 @@ function get_dynamic_blocks_regex() {
  * Renders a single block into a HTML string.
  *
  * @since 1.9.0
+ * @since 4.4.0 renders full nested tree of blocks before reassembling into HTML string
+ * @global WP_Post $post The post to edit.
  *
  * @param  array $block A single parsed block object.
  * @return string String of rendered HTML.
  */
 function gutenberg_render_block( $block ) {
-	$block_name  = isset( $block['blockName'] ) ? $block['blockName'] : null;
-	$attributes  = is_array( $block['attrs'] ) ? $block['attrs'] : array();
-	$raw_content = isset( $block['innerHTML'] ) ? $block['innerHTML'] : null;
+	global $post;
 
-	if ( $block_name ) {
-		$block_type = WP_Block_Type_Registry::get_instance()->get_registered( $block_name );
-		if ( null !== $block_type && $block_type->is_dynamic() ) {
-			return $block_type->render( $attributes );
-		}
+	$block_type = WP_Block_Type_Registry::get_instance()->get_registered( $block['blockName'] );
+	$is_dynamic = $block['blockName'] && null !== $block_type && $block_type->is_dynamic();
+	$has_blocks = ! empty( $block['innerBlocks'] );
+
+	if ( $is_dynamic ) {
+		$attributes  = is_array( $block['attrs'] ) ? (array) $block['attrs'] : array();
+		$global_post = $post;
+		$output      = $block_type->render( $attributes, $block['innerHTML'] );
+		$post        = $global_post;
+
+		return $output;
 	}
 
-	if ( $raw_content ) {
-		return $raw_content;
+	if ( ! $has_blocks ) {
+		return $block['innerHTML'];
 	}
 
-	return '';
+	$output = '';
+	$index  = 0;
+	foreach ( $block['innerContent'] as $chunk ) {
+		$output .= is_string( $chunk ) ? $chunk : gutenberg_render_block( $block['innerBlocks'][ $index++ ] );
+	}
+
+	return $output;
 }
 
 if ( ! function_exists( 'do_blocks' ) ) {
 	/**
 	 * Parses dynamic blocks out of `post_content` and re-renders them.
 	 *
 	 * @since 0.1.0
-	 * @global WP_Post $post The post to edit.
+	 * @since 4.4.0 performs full parse on input post content
 	 *
 	 * @param  string $content Post content.
 	 * @return string          Updated post content.
 	 */
 	function do_blocks( $content ) {
-		global $post;
-
-		$rendered_content      = '';
-		$dynamic_block_pattern = get_dynamic_blocks_regex();
-
-		/*
-		 * Back up global post, to restore after render callback.
-		 * Allows callbacks to run new WP_Query instances without breaking the global post.
-		 */
-		$global_post = $post;
-
-		while ( preg_match( $dynamic_block_pattern, $content, $block_match, PREG_OFFSET_CAPTURE ) ) {
-			$opening_tag     = $block_match[0][0];
-			$offset          = $block_match[0][1];
-			$block_name      = $block_match[1][0];
-			$is_self_closing = isset( $block_match[4] );
-
-			// Reset attributes JSON to prevent scope bleed from last iteration.
-			$block_attributes_json = null;
-			if ( isset( $block_match[3] ) ) {
-				$block_attributes_json = $block_match[3][0];
-			}
+		$without_trailing_newlines = preg_replace( '/(<!--\s+\/?wp:.*?-->)\r?\n?/m', '$1', $content );
```

Comment:
With prior comments in mind, I'm wondering: Do we really need this? Can we feed `$content` directly to `gutenberg_parse_blocks`?

Reply:
there's no technical ned to strip the newlines - mainly I was trying to preserve existing behavior but I don't see why they are needed. if we skip it altogether then the output parse will change but probably not in a semantically-altering way

of note: the existing system only strips the trailing newlines from blocks which aren't registered as dynamic blocks at the time of render, thus making it pretty confusing to match behaviors.

are you in favor of removing all the newline-stripping and update the tests instead?

Reply:
>are you in favor of removing all the newline-stripping and update the tests instead?

This sounds fine, yes.